File: wear/src/main/java/com/codegy/aerlink/services/aerlink/reminders/CalendarsActivity.java
package com.codegy.aerlink.services.aerlink.reminders;

import android.content.Intent;
import android.os.Bundle;
import android.support.wearable.view.ProgressSpinner;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;
import com.codegy.aerlink.Constants;
import com.codegy.aerlink.R;
import com.codegy.aerlink.utils.AerlinkActivity;

import java.util.List;

public class CalendarsActivity extends AerlinkActivity implements ReminderServiceHandler.CalendarsCallback, AdapterView.OnItemClickListener {

    private static final String TAG_LOG = CalendarsActivity.class.getSimpleName();

    private ListView mListView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_calendars);
        setAmbientEnabled();


        mDisconnectedLayout = findViewById(R.id.disconnectedLinearLayout);
        mConnectionInfoTextView = (TextView) findViewById(R.id.connectionInfoTextView);

        mConnectionErrorLayout = findViewById(R.id.connectionErrorLinearLayout);

        mLoadingLayout = findViewById(R.id.loadingLayout);
        mLoadingSpinner = (ProgressSpinner) findViewById(R.id.loadingSpinner);

        mListView = (ListView) findViewById(R.id.listView);
        mListView.setAdapter(new CalendarsListAdapter(CalendarsActivity.this, null));
        mListView.setOnItemClickListener(CalendarsActivity.this);

                /*
                SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(CalendarsActivity.this);
                String calendarsData = sp.getString(Constants.SPK_REMINDER_CALENDARS_DATA, null);
                if (calendarsData != null) {
                    onCalendarsUpdated(calendarsData);
                }
                */

        updateInterface();
    }

    @Override
    public void onEnterAmbient(Bundle ambientDetails) {
        super.onEnterAmbient(ambientDetails);

        ((CalendarsListAdapter) mListView.getAdapter()).setAmbient(true);
    }

    @Override
    public void onExitAmbient() {
        super.onExitAmbient();

        ((CalendarsListAdapter) mListView.getAdapter()).setAmbient(false);
    }

    @Override
    public void onConnectedToDevice() {
        ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

        if (serviceHandler != null) {
            setLoading(true);
            scheduleTimeOutTask();

            serviceHandler.setCalendarsCallback(this);
            serviceHandler.requestCalendarsUpdate(false, new Runnable() {
                @Override
                public void run() {
                    cancelTimeOutTask();
                    setLoading(false);

                    showErrorInterface(true);
                }
            });
        }
        else {
            showErrorInterface(true);
        }
    }

    @Override
    public void onDisconnectedFromDevice() {
        super.onDisconnectedFromDevice();

        ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

        if (serviceHandler != null) {
            serviceHandler.setCalendarsCallback(null);
        }
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

        if (serviceHandler != null) {
            ReminderCalendar calendar = ((CalendarsListAdapter) mListView.getAdapter()).getItem(position);
            serviceHandler.setSelectedCalendar(calendar);

            Intent remindersIntent = new Intent(this, RemindersActivity.class);
            remindersIntent.putExtra(Constants.IE_CALENDAR, calendar);

            startActivity(remindersIntent);
        }
        else {
            onDisconnectedFromDevice();
        }
    }

    @Override
    public void onDataTransferStarted() {
        cancelTimeOutTask();
    }

    @Override
    public void onCalendarsUpdated(final List<ReminderCalendar>  calendars) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                setLoading(false);
                showErrorInterface(false);
                cancelTimeOutTask();

                if (mListView != null) {
                    ((CalendarsListAdapter) mListView.getAdapter()).refresh(calendars);
                }
            }
        });
    }
}

File: wear/src/main/java/com/codegy/aerlink/ui/reminders/CalendarsActivity.kt
package com.codegy.aerlink.ui.reminders

import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.codegy.aerlink.R
import com.codegy.aerlink.service.aerlink.reminders.RemindersServiceManager
import com.codegy.aerlink.service.aerlink.reminders.model.ReminderCalendar
import com.codegy.aerlink.ui.AerlinkActivity
import kotlinx.android.synthetic.main.activity_calendars.*
import kotlinx.android.synthetic.main.list_item_calendar.view.*

class CalendarsActivity : AerlinkActivity(), RemindersServiceManager.CalendarsCallback {
    inner class Adapter: RecyclerView.Adapter<Adapter.ViewHolder>() {
        private var calendars: List<ReminderCalendar> = listOf()

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val layoutInflater: LayoutInflater = LayoutInflater.from(parent.context)
            val itemView = layoutInflater.inflate(R.layout.list_item_calendar, parent, false)
            return ViewHolder(itemView)
        }

        override fun getItemCount(): Int = calendars.size

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val calendar = calendars[position]
            holder.textView.text = calendar.title
            if (isAmbient) {
                holder.textView.setTextColor(Color.LTGRAY)
                holder.cardView.setCardBackgroundColor(Color.rgb(24, 24, 24))
            } else {
                holder.textView.setTextColor(Color.WHITE)
                holder.cardView.setCardBackgroundColor(Color.parseColor("#52" + calendar.color))
            }
            holder.itemView.setOnClickListener {
                RemindersActivity.startWithCalendar(holder.itemView.context, calendar)
            }
        }

        fun reloadData(calendars: List<ReminderCalendar>) {
            this.calendars = calendars
            notifyDataSetChanged()
        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val cardView: CardView = itemView.cardView
            val textView: TextView = itemView.textView
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_calendars)
        setAmbientEnabled()

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = Adapter()
    }

    override fun onPause() {
        val serviceManager = getServiceManager(RemindersServiceManager::class)
        serviceManager?.calendarsCallback = null
        super.onPause()
    }

    override fun onEnterAmbient(ambientDetails: Bundle?) {
        super.onEnterAmbient(ambientDetails)
        recyclerView.adapter?.notifyDataSetChanged()
    }

    override fun onExitAmbient() {
        super.onExitAmbient()
        recyclerView.adapter?.notifyDataSetChanged()
    }

    override fun onConnectedToDevice() {
        val serviceManager = getServiceManager(RemindersServiceManager::class)
        if (serviceManager == null) {
            Log.v(LOG_TAG, "Service manager not available")
            showError(true)
            return
        }

        serviceManager.calendarsCallback = this
        showLoading(true)
        serviceManager.requestCalendarsUpdate(false)
    }

    override fun onError() {
        showError(true)
    }

    override fun onDataTransferStarted() {
        showLoading(true)
    }

    override fun onCalendarsUpdated(calendars: List<ReminderCalendar>) {
        runOnUiThread {
            showLoading(false)
            val adapter = recyclerView.adapter as? Adapter
            adapter?.reloadData(calendars)
        }
    }

    companion object {
        private val LOG_TAG = CalendarsActivity::class.java.simpleName
    }
}
